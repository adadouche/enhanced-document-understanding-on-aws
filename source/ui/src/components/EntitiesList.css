.ObjectDisplay {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }
  
  .ColumnHeader {
    font-weight: bold;
  }
  
  .ObjectRow {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }
  
  .ObjectCell {
    margin: 5px;
  }
  