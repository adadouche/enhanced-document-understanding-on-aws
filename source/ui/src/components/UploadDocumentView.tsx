/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import {
    Alert,
    AppLayout,
    Button,
    Container,
    ContentLayout,
    Form,
    FormField,
    Header,
    ProgressBar,
    Select,
    SelectProps,
    SpaceBetween
} from '@cloudscape-design/components';
import { generateToken, getUsername } from './DocumentTable/DocumentTable';
import { FileSize, FileType } from './FileUpload/interfaces';

import { API } from 'aws-amplify';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import { API_NAME, MAX_UPLOAD_FILE_SIZE } from '../utils/constants';
import FileUpload from './FileUpload/FileUpload';
import { mapResultsToCases } from './makeData';
import { formatFileSize } from './FileUpload/internal';

export interface WorkflowConfig {
    NumRequiredDocuments: number;
    UniqueDocumentTypes: string[];
    WorkflowConfigName: string;
}

type UploadDocumentViewProps = {
    caseName?: string;
    caseId?: string;
    casesList: any[];
    setCasesList: Function;
    workflowConfig: WorkflowConfig;
    selectedCaseRemainingRequiredDocs: any;
};

type Case = {
    name: string;
    status: string;
    caseDocuments: Object[];
    dateCreated: string;
};

type CaseMap = {
    [key: string]: Case;
};

let token: string;
let results: any;

enum Status {
    IN_PROGRESS = 'in-progress',
    SUCCESS = 'success',
    ERROR = 'error'
}

export default function UploadDocumentView(props: UploadDocumentViewProps) {
    let caseMap: CaseMap = {};
    let casesOptions: SelectProps.Option[] = [];

    const updateCaseMap = () => {
        if (props.casesList) {
            for (const item of props.casesList) {
                const _case: Case = {
                    name: item.name,
                    status: item.status,
                    caseDocuments: item.caseDocuments,
                    dateCreated: item.dateCreated
                };
                caseMap[item.caseId] = _case;
            }
        }
        for (const caseId in caseMap) {
            casesOptions.push({
                label: caseMap[caseId].name,
                value: caseId,
                description: `Case Id: ${caseId}`,
                disabled:
                    caseMap[caseId].status !== 'initiate' ||
                    caseMap[caseId].caseDocuments.length >= props.workflowConfig.NumRequiredDocuments
            });
        }
    };

    updateCaseMap();

    let caseId: string | undefined;
    if (props.caseId) {
        caseId = props.caseId;
    }

    const [initialSelectedCaseId] = React.useState<string>(caseId!);
    const [initialSelectedCaseName] = React.useState<string>(caseMap[initialSelectedCaseId]?.name);
    const [selectedCaseOption, setSelectedCaseOption] = React.useState<SelectProps.Option>({
        label: initialSelectedCaseName,
        value: initialSelectedCaseId
    });
    const [chosenFiles, setChosenFiles] = React.useState<File[]>([]);
    const [showProgressBar, setShowProgressBar] = React.useState(false);
    const [progressStatus, setProgressStatus] = React.useState<Status | undefined>();
    const [filesFailedToUpload, setFilesFailedToUpload] = React.useState<string[]>([]);
    const [uploadResultMessage, setUploadResultMessage] = React.useState('');
    const [uploadProgress, setUploadProgress] = React.useState(0);
    const [chooseFilesError, setChooseFilesError] = React.useState('');
    const [allFilesAttempted, setAllFilesAttempted] = React.useState(false);
    const [documentTypeList, setDocumentTypeList] = React.useState<string[]>([]);
    const [refreshCaseData, setRefreshCaseData] = React.useState(false);
    const [selectedCaseRemainingAllowedDocs, setSelectedCaseRemainingAllowedDocs] = React.useState();
    const navigate = useNavigate();

    const fileMetadata = { name: true, type: true, size: FileSize.MB, lastModified: true, thumbnail: true };
    const allowMultipleFiles = true;

    React.useEffect(() => {
        const newRemainingDocs = Object.assign({}, props.selectedCaseRemainingRequiredDocs);

        for (let docType of documentTypeList) {
            if (newRemainingDocs[docType]) {
                newRemainingDocs[docType] -= 1;
            }
        }
        setSelectedCaseRemainingAllowedDocs(newRemainingDocs);
    }, [documentTypeList, props.selectedCaseRemainingRequiredDocs]);

    React.useEffect(() => {
        const fetchData = async () => {
            token = await generateToken();
            results = await API.get(API_NAME, 'cases', {
                headers: {
                    Authorization: token
                }
            });
            const newData = mapResultsToCases(results.Items);
            props.setCasesList(newData);
        };
        if (!props.casesList || refreshCaseData) {
            fetchData();
            updateCaseMap();
            setRefreshCaseData(false);
        }
    }, [refreshCaseData]);

    const handleFileSelection = async (newFiles: FileType) => {
        setChooseFilesError('');
        if (newFiles instanceof File) {
            if (MAX_UPLOAD_FILE_SIZE && newFiles.size > MAX_UPLOAD_FILE_SIZE) {
                setChooseFilesError(
                    'File size must be less than ' + formatFileSize(MAX_UPLOAD_FILE_SIZE, fileMetadata)
                );
            } else {
                setChosenFiles([newFiles]);
            }
        } else if (newFiles instanceof Array) {
            if (MAX_UPLOAD_FILE_SIZE) {
                const numFilesAttempted = newFiles.length;
                newFiles = newFiles.filter((file) => file.size <= MAX_UPLOAD_FILE_SIZE);
                if (newFiles.length < numFilesAttempted) {
                    setChooseFilesError(
                        'File size must be less than ' + formatFileSize(MAX_UPLOAD_FILE_SIZE, fileMetadata)
                    );
                }
            }
            setChosenFiles(newFiles);
        }
    };

    const uploadFile = async (
        files: File[],
        index: number,
        postPolicyFields: { [key: string]: string },
        bucketUrl: string
    ) => {
        let postData = new FormData();
        for (const [key, entry] of Object.entries(postPolicyFields)) {
            postData.append(key, entry);
        }
        postData.append('file', files[index]);
        const data = {
            method: 'POST',
            body: postData
        };
        const response = await fetch(bucketUrl, data);
        if (200 <= response.status && response.status < 300) {
            setUploadProgress((uploadProgress) => uploadProgress + Math.ceil((1 / chosenFiles.length) * 100));
        } else {
            setFilesFailedToUpload((filesFailedToUpload) => [...filesFailedToUpload, files[index].name]);
        }
        if (index === chosenFiles.length - 1) {
            setAllFilesAttempted(true);
        }
    };

    const postResource = async (endpoint: string, params = {}) => {
        try {
            token = await generateToken();
            const response = await API.post(API_NAME, endpoint, {
                headers: {
                    Authorization: token
                },
                body: params
            });
            return response;
        } catch (err) {
            console.error(err);
        }
    };

    const handleUploadClick = async () => {
        setProgressStatus(Status.IN_PROGRESS);
        setFilesFailedToUpload((filesFailedToUpload) => []);
        setUploadResultMessage('');
        setUploadProgress(0);
        setShowProgressBar(true);
        setAllFilesAttempted(false);

        const userName = await getUsername();
        let promiseArray = [];
        for (let i = 0; i < chosenFiles.length; i++) {
            let apiRequest = postResource(`document`, {
                userId: userName,
                caseId: selectedCaseOption.value,
                caseName: selectedCaseOption.label,
                fileName: chosenFiles[i].name,
                fileExtension: '.' + chosenFiles[i].name.split('.').pop(),
                documentType: documentTypeList[i],
                tagging: `<Tagging><TagSet><Tag><Key>userId</Key><Value>${userName}</Value></Tag></TagSet></Tagging>`
            });
            promiseArray.push(apiRequest);
        }

        await Promise.allSettled(promiseArray).then(async (results) => {
            results.forEach((result, i) => {
                if (result.status === 'fulfilled') {
                    if (result.value === undefined) {
                        setFilesFailedToUpload((filesFailedToUpload) => [...filesFailedToUpload, chosenFiles[i].name]);
                        setUploadResultMessage((uploadResultMessage) => 'File upload failed');
                        setProgressStatus(Status.ERROR);
                    } else {
                        uploadFile(chosenFiles, i, result.value.fields, result.value.url);
                    }
                }
            });
        });
        setRefreshCaseData(true);
        setChosenFiles([]);
    };

    const allowChooseFiles = () => {
        if (selectedCaseOption.value !== undefined && caseMap[selectedCaseOption.value] !== undefined) {
            return (
                chosenFiles.length <
                props.workflowConfig.NumRequiredDocuments - caseMap[selectedCaseOption.value].caseDocuments.length
            );
        }
        return false;
    };

    const updateDocumentTypeList = (index: number, value: string | null) => {
        const updatedList = [...documentTypeList];
        if (index >= updatedList.length) {
            updatedList.push(value!);
        } else if (value === null) {
            updatedList.splice(index, 1);
        } else {
            updatedList[index] = value;
        }
        setDocumentTypeList(updatedList);
    };

    const allowUploadFiles = () => {
        if (selectedCaseOption.value !== undefined && caseMap[selectedCaseOption.value] !== undefined) {
            return (
                chosenFiles.length > 0 &&
                caseMap[selectedCaseOption.value].caseDocuments.length + chosenFiles.length <=
                    props.workflowConfig.NumRequiredDocuments
            );
        }
        return false;
    };

    React.useEffect(() => {
        if (uploadProgress < 100 && allFilesAttempted) {
            setProgressStatus(Status.ERROR);
            setUploadResultMessage((uploadResultMessage) => 'File upload failed');
        } else if (uploadProgress >= 100 && allFilesAttempted) {
            setProgressStatus(Status.SUCCESS);
            setUploadResultMessage((uploadResultMessage) => 'Files uploaded sucessfully');
        } else if (uploadProgress > 0 && uploadProgress < 100 && !allFilesAttempted) {
            setProgressStatus(Status.IN_PROGRESS);
        }
    }, [uploadProgress, progressStatus, allFilesAttempted]);

    return (
        <div>
            <AppLayout
                contentType="form"
                content={
                    <ContentLayout
                        header={
                            <SpaceBetween size="m">
                                <Header variant="h1">Upload Documents</Header>
                            </SpaceBetween>
                        }
                        data-testid="doc-upload-view-contentlayout"
                    >
                        <form onSubmit={(event) => event.preventDefault()}>
                            <Form
                                actions={
                                    <SpaceBetween direction="horizontal" size="xs">
                                        {showProgressBar && (
                                            <ProgressBar
                                                value={uploadProgress}
                                                additionalInfo={uploadResultMessage}
                                                variant="key-value"
                                                label="Document Upload Status"
                                                status={progressStatus}
                                                data-testid="upload-document-progress-bar"
                                            />
                                        )}
                                        <span>&nbsp;&nbsp;&nbsp;&nbsp;</span>
                                        <Button
                                            data-testid="upload-documents-button"
                                            variant="primary"
                                            onClick={() => handleUploadClick()}
                                            disabled={!allowUploadFiles()}
                                        >
                                            Upload Documents
                                        </Button>
                                        <Button variant="link" onClick={() => navigate(`/`)}>
                                            Done
                                        </Button>
                                    </SpaceBetween>
                                }
                                errorIconAriaLabel="Error"
                            >
                                <Container>
                                    <SpaceBetween direction="vertical" size="l">
                                        <FormField
                                            label="Case name"
                                            i18nStrings={{ errorIconAriaLabel: 'Error' }}
                                            data-testid="case-name-field"
                                        >
                                            <Select
                                                selectedOption={selectedCaseOption}
                                                onChange={({ detail }) => setSelectedCaseOption(detail.selectedOption)}
                                                options={casesOptions}
                                                filteringType="auto"
                                                selectedAriaLabel="Selected"
                                                data-testid="upload-doc-case-select"
                                            />
                                        </FormField>
                                        <FileUpload
                                            label="Files to upload"
                                            accept=".png,.pdf,.jpg,.jpeg"
                                            fileMetadata={fileMetadata}
                                            multiple={allowMultipleFiles}
                                            value={chosenFiles}
                                            description={
                                                'Please choose the files you wish to upload to the selected case'
                                            }
                                            disabled={!allowChooseFiles()}
                                            errorText={chooseFilesError}
                                            data-testid="file-upload-field"
                                            onChange={({ detail }) => handleFileSelection(detail.value)}
                                            workflowConfig={props.workflowConfig}
                                            documentTypeList={documentTypeList}
                                            setDocumentTypeList={updateDocumentTypeList}
                                            remainingAllowedDocs={selectedCaseRemainingAllowedDocs}
                                        ></FileUpload>
                                    </SpaceBetween>
                                </Container>
                            </Form>
                        </form>
                        {uploadResultMessage === 'File upload failed' && filesFailedToUpload.length > 0 && (
                            <Alert
                                statusIconAriaLabel="Error"
                                type="error"
                                header="Failed to upload the following files: "
                            >
                                {filesFailedToUpload.map((fileName, i) => (
                                    <li key={fileName + i}>{fileName}</li> // NOSONAR - fix to make displayed file names unique to be implemented in the future
                                ))}
                            </Alert>
                        )}
                    </ContentLayout>
                }
                headerSelector="#header"
                navigationHide
                toolsHide
                date-testid="doc-upload-view-applayout"
            />
        </div>
    );
}
