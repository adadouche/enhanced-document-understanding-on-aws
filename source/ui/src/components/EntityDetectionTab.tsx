/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Box, Container, StatusIndicatorProps } from '@cloudscape-design/components';
import { useCallback, useMemo } from 'react';
import { renderStatus } from '../utils/common-renderers';
import { BoundingBox } from '../utils/interfaces';

import { EntityTypes } from '../utils/constants';
import DocumentRenderer from './DocumentRenderer/DocumentRenderer';
import EntitiesList from './EntitiesList';
import { getEntitiesToProcess } from './entityUtils';

type EntityDetectionTabProps = {
    selectedDocumentFileType: string | null;
    selectedDocumentUrl: string | null;
    selected?: string;
    standardEntities: any;
    medicalEntities: any;
    piiEntities: any;
    documentPageCount: number;
    currentPageNumber: number;
    switchPage: Function;
    comprehendService: string;
    entityType: string;
    selectedDocumentId: string | null;
    selectedCaseId: string | null;
    selectedEntities: any;
    setSelectedEntities: Function;
    previewRedaction: string;
    setPreviewRedaction: Function;
    currentStatus: StatusIndicatorProps.Type | undefined;
    dataTestId?: string;
    retrieveSignedUrl: Function;
};

/**
 * EntityDetectionTab is used to render data for Generic, PII and Medical Comprehend inferences
 * @param props to be used to populate the tab
 * @returns
 */
export default function EntityDetectionTab(props: EntityDetectionTabProps) {
    let documentEntities = getEntitiesToProcess(props.entityType, props);

    const getFilteredArray = useCallback(
        (entityType: string) => {
            return props.selectedEntities[entityType].filter((item: string[]) => {
                if (item.length > 2 && item[2] !== props.currentPageNumber.toString()) {
                    return false;
                }

                if (
                    props.selectedEntities[entityType].some(
                        (otherItem: string[]) =>
                            otherItem.length < item.length &&
                            otherItem.slice(0, item.length - 1).every((val, index) => val === item[index])
                    )
                ) {
                    return false;
                }

                return true;
            });
        },
        [props.currentPageNumber, props.selectedEntities]
    );

    const getBoundingBoxesForEntityPage = useCallback((entities: any, entityPath: string[]) => {
        let boundingBoxes: BoundingBox[] = [];
        for (const pageInstance of entities[entityPath[0]][entityPath[1]][entityPath[2]]) {
            boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
        }
        return boundingBoxes;
    }, []);

    const getBoundingBoxesForEntityValue = useCallback(
        (entities: any, entityPath: string[]) => {
            let boundingBoxes: BoundingBox[] = [];
            if (props.currentPageNumber.toString() in entities[entityPath[0]][entityPath[1]]) {
                for (const pageInstance of entities[entityPath[0]][entityPath[1]][props.currentPageNumber.toString()]) {
                    boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
                }
            }
            return boundingBoxes;
        },
        [props.currentPageNumber]
    );

    const getBoundingBoxesForEntityType = useCallback(
        (entities: any, entityPath: string[]) => {
            let boundingBoxes: BoundingBox[] = [];
            for (const entity of Object.keys(entities[entityPath[0]])) {
                if (props.currentPageNumber.toString() in entities[entityPath[0]][entity]) {
                    for (const pageInstance of entities[entityPath[0]][entity][props.currentPageNumber.toString()]) {
                        boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
                    }
                }
            }
            return boundingBoxes;
        },
        [props.currentPageNumber]
    );

    const pageEntities = useMemo(() => {
        let boundingBoxes: BoundingBox[] = [];
        for (const entityType of Object.values(EntityTypes)) {
            if (props.selectedEntities[entityType] && props.selectedEntities[entityType].length > 0) {
                const entities = getEntitiesToProcess(entityType, props);
                const filteredArray = getFilteredArray(entityType);
                for (const entityPath of filteredArray) {
                    switch (entityPath.length) {
                        case 3:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityPage(entities, entityPath));
                            break;
                        case 2:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityValue(entities, entityPath));
                            break;
                        case 1:
                        default:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityType(entities, entityPath));
                            break;
                    }
                }
            }
        }

        return boundingBoxes;
    }, [
        getBoundingBoxesForEntityPage,
        getBoundingBoxesForEntityType,
        getBoundingBoxesForEntityValue,
        getFilteredArray,
        props.medicalEntities,
        props.piiEntities,
        props.selectedEntities,
        props.standardEntities
    ]);

    const status = renderStatus(
        props.currentStatus,
        true,
        false,
        `An error occurred loading the document preview.`,
        ''
    );
    return (
        <div style={{ display: 'flex', height: '100%' }} data-testid={props.dataTestId}>
            <div style={{ width: '50%', float: 'left', paddingRight: '0.5%', paddingLeft: '1%' }}>
                <Container data-testid="document-rendering-container">
                    {status}
                    <DocumentRenderer
                        selectedDocumentFileType={props.selectedDocumentFileType}
                        selectedDocumentUrl={props.selectedDocumentUrl}
                        currentPageNumber={props.currentPageNumber}
                        switchPage={props.switchPage}
                        marks={pageEntities}
                        previewRedaction={props.previewRedaction}
                        retrieveSignedUrl={props.retrieveSignedUrl}
                    />
                </Container>
            </div>
            <div
                style={{
                    width: '50%',
                    float: 'left',
                    paddingLeft: '0.5%',
                    paddingRight: '1%',
                    height: '100%'
                }}
            >
                <Container fitHeight={true}>
                    <div>
                        <Box data-testid="tab-box">
                            <EntitiesList
                                entities={documentEntities}
                                documentPageCount={props.documentPageCount}
                                currentPageNumber={props.currentPageNumber}
                                switchPage={props.switchPage}
                                comprehendService={props.comprehendService}
                                entityType={props.entityType}
                                standardEntities={props.standardEntities}
                                medicalEntities={props.medicalEntities}
                                piiEntities={props.piiEntities}
                                selectedEntities={props.selectedEntities}
                                setSelectedEntities={props.setSelectedEntities}
                                selectedDocumentId={props.selectedDocumentId}
                                selectedCaseId={props.selectedCaseId}
                                previewRedaction={props.previewRedaction}
                                setPreviewRedaction={props.setPreviewRedaction}
                                currentStatus={props.currentStatus}
                            />
                        </Box>
                    </div>
                </Container>
            </div>
        </div>
    );
}
