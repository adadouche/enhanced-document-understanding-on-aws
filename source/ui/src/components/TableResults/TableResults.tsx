/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import { Box, ExpandableSection, StatusIndicatorProps } from '@cloudscape-design/components';
import { Fragment, useEffect, useRef } from 'react';
import { isStatusSuccess, renderStatus } from '../../utils/common-renderers';
import { TextractTable } from '../../utils/interfaces';
import './TableResults.css';

type TableResultsProps = {
    tables: TextractTable[];
    documentPageCount: number;
    currentPageNumber: number;
    switchPage: Function;
    currentStatus: StatusIndicatorProps.Type | undefined;
};

export default function TableResults(props: TableResultsProps) {
    const container = useRef<HTMLDivElement>(null);

    useEffect(() => {
        if (container.current) {
            const firstOnThisPage = container.current.querySelector('raw_text_lines_' + props.currentPageNumber);
            if (firstOnThisPage) firstOnThisPage.scrollIntoView();
        }
    }, [props.currentPageNumber]);

    const status = renderStatus(props.currentStatus, true, false, 'An error occurred loading table results.', '');

    // If successful request, but no inferences received.
    if (isStatusSuccess(props.currentStatus) && !props.tables.length) {
        return <p data-testid="tables-nodata">No Tables detected</p>;
    } else {
        // successful and data exists / loading the results still / there is an error
        if (props.tables.length > 0) {
            const groupedTables = groupPagesByPage(props.tables);
            return (
                <div ref={container}>
                    {status}
                    <Box variant="h3" display="inline">
                        Tables: {props.tables.length || 0} Found
                    </Box>
                    <ul data-testid="tables-with-data">
                        {groupedTables.map((tables: TextractTable[], i: number) => (
                            <Fragment key={tables[0].pageNumber}>
                                <ExpandableSection headerText={'Page ' + tables[0].pageNumber}>
                                    {tables.map(({ table, pageNumber, rows }, i) => (
                                        <div
                                            key={tables[0].pageNumber + JSON.stringify(rows)}
                                            onClick={() => {
                                                props.switchPage(pageNumber);
                                            }}
                                            style={{ cursor: 'pointer' }}
                                        >
                                            <h5>
                                                Table {i + 1} : {rows.length || 0} rows
                                            </h5>
                                        </div>
                                    ))}
                                </ExpandableSection>
                            </Fragment>
                        ))}
                    </ul>
                    <footer />
                </div>
            );
        }
        return <p data-testid="tables-status-only"> {status} </p>; // data still loading / error occurred
    }
}

const groupPagesByPage = (tables: TextractTable[]) => {
    const groupedTables: TextractTable[][] = [];
    let currentPage: number | null = null;
    let currentGroup: TextractTable[] = [];
    for (const table of tables) {
        if (currentPage !== table.pageNumber) {
            if (currentPage !== null) {
                groupedTables.push(currentGroup);
            }
            currentPage = table.pageNumber;
            currentGroup = [table];
        } else {
            currentGroup.push(table);
        }
    }
    groupedTables.push(currentGroup);
    return groupedTables;
};
