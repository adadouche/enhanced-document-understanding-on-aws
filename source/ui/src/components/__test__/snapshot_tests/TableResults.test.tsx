/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { StatusIndicatorProps } from '@cloudscape-design/components';
import { tables } from '../../../__test__/test_data';
import TableResults from '../../TableResults/TableResults';

const tableResultsPropsEmpty = {
    currentPageNumber: 1,
    documentPageCount: 4,
    tables: [],
    switchPage: jest.fn(),
    currentStatus: 'success' as StatusIndicatorProps.Type
};

const tableResultsProps = {
    currentPageNumber: 1,
    documentPageCount: 4,
    tables: tables,
    switchPage: jest.fn(),
    currentStatus: 'success' as StatusIndicatorProps.Type
};

const tableResultsPropsLoading = {
    currentPageNumber: 0,
    documentPageCount: 0,
    tables: [],
    switchPage: jest.fn(),
    currentStatus: 'loading' as StatusIndicatorProps.Type
};

const tableResultsPropsFailed = {
    currentPageNumber: 0,
    documentPageCount: 0,
    tables: [],
    switchPage: jest.fn(),
    currentStatus: 'error' as StatusIndicatorProps.Type
};

jest.mock('@cloudscape-design/components', () => {
    const Components = jest.genMockFromModule('@cloudscape-design/components') as any;
    for (const componentName of Object.keys(Components)) {
        Components[componentName] = componentName;
    }
    return Components;
});

test('Renders without tables', async () => {
    const tree = renderer.create(<TableResults {...tableResultsPropsEmpty} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('Renders error message when currentStatus is error', async () => {
    const tree = renderer.create(<TableResults {...tableResultsPropsFailed} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('Renders loading message when currentStatus is loading', async () => {
    const tree = renderer.create(<TableResults {...tableResultsPropsLoading} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('Renders tables with page numbers shown for multi-page', async () => {
    const tree = renderer.create(<TableResults {...tableResultsProps} />).toJSON();
    expect(tree).toMatchSnapshot();
});
