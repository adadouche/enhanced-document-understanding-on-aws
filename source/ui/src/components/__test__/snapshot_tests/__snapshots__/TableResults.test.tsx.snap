// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders error message when currentStatus is error 1`] = `
<p
  data-testid="tables-status-only"
>
   
  <span
    className="awsui_root_1cbgc_1u3x7_147 awsui_status-error_1cbgc_1u3x7_156"
  >
    <span
      className="awsui_container_1cbgc_1u3x7_125"
    >
      <span
        className="awsui_icon_1cbgc_1u3x7_97"
      >
        <span
          className="awsui_icon_h11ix_ttpon_98 awsui_size-normal-mapped-height_h11ix_ttpon_151 awsui_size-normal_h11ix_ttpon_147 awsui_variant-normal_h11ix_ttpon_219"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 16 16\\" focusable=\\"false\\" aria-hidden=\\"true\\"><circle class=\\"stroke-linejoin-round\\" cx=\\"8\\" cy=\\"8\\" r=\\"7\\"/><path d=\\"m10.828 5.172-5.656 5.656M10.828 10.828 5.172 5.172\\"/></svg>",
            }
          }
          style={Object {}}
        />
      </span>
      An error occurred loading table results.
    </span>
  </span>
   
</p>
`;

exports[`Renders loading message when currentStatus is loading 1`] = `
<p
  data-testid="tables-status-only"
>
   
  <span
    className="awsui_root_1cbgc_1u3x7_147 awsui_status-loading_1cbgc_1u3x7_177"
  >
    <span
      className="awsui_container_1cbgc_1u3x7_125"
    >
      <span
        className="awsui_icon_1cbgc_1u3x7_97"
      >
        <span
          className="awsui_root_1612d_qodmh_127 awsui_size-normal_1612d_qodmh_135 awsui_variant-normal_1612d_qodmh_159"
        >
          <span
            className="awsui_circle_1612d_qodmh_169 awsui_circle-left_1612d_qodmh_193"
          />
          <span
            className="awsui_circle_1612d_qodmh_169 awsui_circle-right_1612d_qodmh_198"
          />
        </span>
      </span>
      Loading
    </span>
  </span>
   
</p>
`;

exports[`Renders tables with page numbers shown for multi-page 1`] = `
<div>
  <Box
    display="inline"
    variant="h3"
  >
    Tables: 
    2
     Found
  </Box>
  <ul
    data-testid="tables-with-data"
  >
    <ExpandableSection
      headerText="Page 1"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <h5>
          Table 
          1
           : 
          2
           rows
        </h5>
      </div>
    </ExpandableSection>
    <ExpandableSection
      headerText="Page 2"
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <h5>
          Table 
          1
           : 
          1
           rows
        </h5>
      </div>
    </ExpandableSection>
  </ul>
  <footer />
</div>
`;

exports[`Renders without tables 1`] = `
<p
  data-testid="tables-nodata"
>
  No Tables detected
</p>
`;
